File: javascript/101-symmetric-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
 var isSymmetric = function(root) {
    if (!root) return true
    const search = (left, right) => {
        if (!left && !right) return true
        if (left && !right || !left && right || left.val !== right.val) return false
        return search(left.right, right.left) && search(left.left, right.right)
    }
    return search(root.left, root.right)
}

File: golang/101-symmetric-tree.go
package golang

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func search(left *TreeNode, right *TreeNode) bool {
    if left == nil && right == nil {
        return true
    }
    if right == nil || left == nil || left.Val != right.Val {
        return false
    }
    return search(left.Right, right.Left) && search(left.Left, right.Right)
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return search(root.Left, root.Right)
}